### ISSUE TITLE
Require access to Scenario and Feature tags in the Velocity custom VM template

### ISSUE THREAD

**Issue by Hyphen31** [temyers/cucumber-jvm-parallel-plugin#112]
We are using the customVmTemplate feature of the plugin but we need to add logic to our template based  on Scenario and Feature tags.

These are not currently available to the Velocity template but I have made the respective changes locally to enable this

If you are willing for me to push my changes to your project then I can create a Pull Request.

**Comment by mpkorstanje** [comment]
I don't see a real problem with making the whole feature (not just the tags) and scenario object available to the template. However I would like to understand the use case a bit more.

**Comment by Hyphen31** [comment]
We are using Cucumber for testing APIs and end-to-end UI tests. We use an @api annotation on our Cucumber tests against the scenario (or sometimes the Feature) to indicate it is an API test and hence we do not need a browser launched. This is controlled by extending a different BaseTest for api and ui tests.

Assuming the Velocity template supplies a variable called ``$scenarioTags`` then our customVmTemplate will be:
```
#parse("/array.java.vm")
#if ($packageName)
package $packageName;

#end##
import cucumber.api.CucumberOptions;
import cucumber.api.testng.AbstractTestNGCucumberTests;
import org.testng.annotations.Test;
import.com.frameworkium.core.common.listeners.CukeListener;
#if( $scenarioTags && $scenarioTags.contains("@api") )
import com.frameworkium.core.api.tests.BaseTest;
#else
import com.frameworkium.core.ui.tests.BaseTest;
#end

public class $className extends BaseTest {
    
    @Test
    @CucumberOptions(
            strict = $strict,
            features = {"$featureFile"},
            plugin = #stringArray($plugins),
            monochrome = ${monochrome},
#if(!$featureFile.contains(".feature:") && $tags)
            tags = #stringArray($tags),
#end
            glue = #stringArray($glue))
    public class test_$className extends AbstractTestNGCucumberTests {}
}
```

**Comment by mpkorstanje** [comment]
Okay. This can be useful.

I would like to see a few changes though. 
 - Please  expose the current Feature object in CucumberITGeneratorByFeature
 - Please  expose the current Feature and current Scenario or ScenarioOutline object in CucumberITGeneratorByScenario

You should be able get the tags directly from those objects.

---

As an aside: could you explain why wrapping your webdriver in a class that doesn't initialize the web-driver (and thus launching the browser) until it is actually first used?

```java
public class LazyWebDriver implements Disposable, org.openqa.selenium.WebDriver, TakesScreenshot {

    private org.openqa.selenium.WebDriver driver;

    @Override
    public void dispose() {
        if(driver == null){
            return;
        }

        driver.quit();
    }

    private org.openqa.selenium.WebDriver getDriver() {
        // TODO: Use maven profile to determine type
        if (driver == null) {
            driver = new FirefoxDriver();
        }
        return driver;
    }

    boolean hasBeenUsed() {
        return driver != null;
    }

    @Override
    public void get(String url) {
        getDriver().get(url);
    }

   ...more delegation...

    @Override
    public Navigation navigate() {
        return getDriver().navigate();
    }

    @Override
    public Options manage() {
        return getDriver().manage();
    }

    @Override
    public <X> X getScreenshotAs(OutputType<X> target) throws WebDriverException {
        WebDriver driver = getDriver();
        assertTrue(driver instanceof  TakesScreenshot);
        return ((TakesScreenshot)driver).getScreenshotAs(target);
    }
}
```

**Comment by Hyphen31** [comment]
OK. Can you please give me permission to submit my changes as a PR.

**Comment by mpkorstanje** [comment]
You shouldn't need permission for that. You can fork the repo and push the changes there. Then use the create pull request button on your fork of the repo. If you've cloned this repo directly you may need to change the origin to point to your fork.

https://help.github.com/articles/changing-a-remote-s-url/
https://help.github.com/articles/fork-a-repo/
https://help.github.com/articles/about-pull-requests/

**Comment by mpkorstanje** [comment]
Resolved with  		1439968  #116

**Comment by temyers** [comment]
Fixed in 4.1.0
